File: aoc.main/src/aoc/day02.clj
(ns aoc.day02
  (:require [aoc.util :as util]
            [clojure.string :as str]))

(defn parse-line [line]
  (let [[direction amount] (str/split line #" ")]
    [direction (Integer. amount)]))

(defn parse-data [input]
  (->> input
       str/split-lines
       (map parse-line)))

(defn solve-part1 []
  (let [input (-> "2" util/get-input parse-data)
        [location depth] (reduce (fn [[location depth] [direction amount]]
                                   (case direction
                                     "forward" [(+ location amount) depth]
                                     "down" [location (+ depth amount)]
                                     "up" [location (- depth amount)]))
                                 [0 0] input)]
    (* location depth)))

(defn solve-part2 []
  (let [input (-> "2" util/get-input parse-data)
        [location depth _] (reduce (fn [[location depth aim] [direction amount]]
                                     (case direction
                                       "forward" [(+ location amount) (+ depth (* aim amount)) aim]
                                       "down" [location depth (+ aim amount)]
                                       "up" [location depth (- aim amount)]))
                                   [0 0 0] input)]
    (* location depth)))

File: src/day02.rs
use aoc_runner_derive::{aoc, aoc_generator};

#[aoc_generator(day2)]
pub fn pares_input(input: &str) -> Vec<(String, i32)> {
    input
        .lines()
        .map(|l| {
            let mut s = l.split(' ');
            (
                s.next().unwrap().to_owned(),
                s.next().unwrap().parse::<i32>().unwrap(),
            )
        })
        .collect()
}

#[aoc(day2, part1)]
pub fn solve_part1(input: &Vec<(String, i32)>) -> i32 {
    let mut depth = 0;
    let mut position = 0;

    for (action, x) in input {
        match action.as_str() {
            "forward" => position += x,
            "down" => depth += x,
            "up" => depth -= x,
            _ => panic!("Unknown direction {}", action),
        }
    }

    depth * position
}

#[aoc(day2, part2)]
pub fn solve_part2(input: &Vec<(String, i32)>) -> i32 {
    let mut depth = 0;
    let mut position = 0;
    let mut aim = 0;

    for (action, x) in input {
        match action.as_str() {
            "forward" => {
                position += x;
                depth += x * aim
            }
            "down" => aim += x,
            "up" => aim -= x,

            _ => panic!("Unknown direction {}", action),
        }
    }

    depth * position
}
